oauthproxy.go

```diff
@@ -761,48 +746,48 @@ func (p *OAuthProxy) OAuthCallback(rw http.ResponseWriter, req *http.Request) {
 	err := req.ParseForm()
 	if err != nil {
 		logger.Errorf("Error while parsing OAuth2 callback: %v", err)
-		p.ErrorPage(rw, req, http.StatusInternalServerError, "Internal Server Error", err.Error())
+		p.ErrorPage(rw, req, http.StatusInternalServerError, err.Error())
 		return
 	}
 	errorString := req.Form.Get("error")
 	if errorString != "" {
 		logger.Errorf("Error while parsing OAuth2 callback: %s", errorString)
-		p.ErrorPage(rw, req, http.StatusForbidden, "Permission Denied", errorString)
+		p.ErrorPage(rw, req, http.StatusForbidden, errorString, "Login Failed: The upstream identity provider returned an error.")
 		return
 	}
 
 	session, err := p.redeemCode(req)
 	if err != nil {
 		logger.Errorf("Error redeeming code during OAuth2 callback: %v", err)
-		p.ErrorPage(rw, req, http.StatusInternalServerError, "Internal Server Error", "Internal Error")
+		p.ErrorPage(rw, req, http.StatusInternalServerError, err.Error(), "Login Failed: There was a problem with the code returned by the upstream identity provider.")
 		return
 	}
 
 	err = p.enrichSessionState(req.Context(), session)
```

Comment:
This `EnrichSessionState` call in general should be after the `state` CSRF protection.

Otherwise even if we align the error messages to be the same to not leak details, this is a noticeable network request in the middle -- so timing details would leak and tip our hand to attackers.